pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +504,113 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) ownedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) getAndClearReasonForRecomputeOwnedSeries() string {
+	return u.requiresOwnedSeriesUpdate.Swap("")
+}
+
+func (u *userTSDB) triggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+// recomputeOwnedSeries recomputes owned series for current token ranges, and updates both owned series and shard size.
+//
+// This method returns true, if recomputation of owned series failed multiple times due to too
+// many new series being added during the computation. If no such problem happened, this method returns false.
+//
+// This method and updateTokenRanges should be only called from the same goroutine. (ownedSeries service)
+func (u *userTSDB) recomputeOwnedSeries(shardSize int, reason string, logger log.Logger) (retry bool) {
+	retry, _ = u.recomputeOwnedSeriesWithComputeFn(shardSize, reason, logger, u.computeOwnedSeries)
+	return retry
+}
+
+const (
+	recomputeOwnedSeriesMaxAttemps    = 3
+	recomputeOwnedSeriesMaxSeriesDiff = 1000
+)
+
+func (u *userTSDB) recomputeOwnedSeriesWithComputeFn(shardSize int, reason string, logger log.Logger, compute func() int) (retry bool, _ int) {
```

Comment:
[nit] It looks a bit weird to me calling the return param `retry` (even in the callers) when it's effectively a `failed`.

Reply:
I don't like calling return value "failed" (I prefer "success"), hence I opted for "retry". I will rename it to `success` instead.